pkg/xds/topology/outbound.go

```diff
@@ -527,19 +535,170 @@ func fillExternalServicesReachableFromZone(
 			createExternalServiceEndpoint(ctx, outbound, externalService, mesh, loader, zone)
 		}
 	}
+	for _, mes := range meshExternalServices {
+		if mes.IsReachableFromZone(zone) {
+			err := createMeshExternalServiceEndpoint(ctx, outbound, mes, mesh, loader, zone)
+			if err != nil {
+				core.Log.Error(err, "unable to create MeshExternalService endpoint. Endpoint won't be included in the XDS.", "name", mes.Meta.GetName(), "mesh", mes.Meta.GetMesh())
+				return
+			}
+		}
+	}
 }
 
 func fillExternalServicesOutbounds(
 	ctx context.Context,
 	outbound core_xds.EndpointMap,
 	externalServices []*core_mesh.ExternalServiceResource,
+	meshExternalServices []*meshexternalservice_api.MeshExternalServiceResource,
 	mesh *core_mesh.MeshResource,
 	loader datasource.Loader,
 	zone string,
 ) {
 	for _, externalService := range externalServices {
 		createExternalServiceEndpoint(ctx, outbound, externalService, mesh, loader, zone)
 	}
+
+	for _, mes := range meshExternalServices {
+		err := createMeshExternalServiceEndpoint(ctx, outbound, mes, mesh, loader, zone)
+		if err != nil {
+			core.Log.Error(err, "unable to create MeshExternalService endpoint. Endpoint won't be included in the XDS.", "name", mes.Meta.GetName(), "mesh", mes.Meta.GetMesh())
+		}
+	}
+}
+
+func createMeshExternalServiceEndpoint(
+	ctx context.Context,
+	outbounds core_xds.EndpointMap,
+	mes *meshexternalservice_api.MeshExternalServiceResource,
+	mesh *core_mesh.MeshResource,
+	loader datasource.Loader,
+	zone string,
+) error {
+	es := &core_xds.ExternalService{
+		Protocol: core_mesh.ParseProtocol(string(mes.Spec.Match.Protocol)),
+	}
+	tags := maps.Clone(mes.Meta.GetLabels())
+	if tags == nil {
+		tags = map[string]string{}
+	}
+	meshName := mesh.GetMeta().GetName()
+	name := mes.Meta.GetName()
+	tls := mes.Spec.Tls
+	if tls != nil && tls.Enabled {
+		var caCert, clientCert, clientKey []byte
+		es.TLSEnabled = tls.Enabled
+		es.FallbackToSystemCa = true
+		es.AllowRenegotiation = tls.AllowRenegotiation
+
+		var err error
+		if tls.Verification != nil {
+			if tls.Verification.CaCert != nil {
+				caCert, err = loadBytes(ctx, tls.Verification.CaCert.ConvertToProto(), meshName, loader)
+				if err != nil {
+					return errors.Wrap(err, "could not load caCert")
+				}
+				es.CaCert = caCert
+			}
+			if tls.Verification.ClientKey != nil && tls.Verification.ClientCert != nil {
+				clientCert, err = loadBytes(ctx, tls.Verification.ClientCert.ConvertToProto(), meshName, loader)
+				if err != nil {
+					return errors.Wrap(err, "could not load clientCert")
+				}
+				clientKey, err = loadBytes(ctx, tls.Verification.ClientKey.ConvertToProto(), meshName, loader)
+				if err != nil {
+					return errors.Wrap(err, "could not load clientKey")
+				}
+				es.ClientCert = clientCert
+				es.ClientKey = clientKey
+			}
+			if tls.Verification.ServerName != nil && pointer.Deref(tls.Verification.ServerName) != "" {
```

Comment:
couldn't it be just: `pointer.Deref(tls.Verification.ServerName) != ""` instead of: tls.Verification.ServerName != nil && pointer.Deref(tls.Verification.ServerName) != ""` Deref will handle nil value

Reply:
yeah, Deref handles nils, I'll change it